monkey/infection_monkey/i_puppet.py

```diff
@@ -0,0 +1,88 @@
+import abc
+import threading
+from collections import namedtuple
+from enum import Enum
+from typing import Dict, Optional, Tuple
+
+
+class PortStatus(Enum):
+    OPEN = 1
+    CLOSED = 2
+
+
+PortScanData = namedtuple("PortScanData", ["port", "status", "banner", "service"])
+
+
+class IPuppet(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def run_sys_info_collector(self, name: str) -> Dict:
+        """
+        Runs a system info collector
+        :param str name: The name of the system info collector to run
+        :return: A dictionary containing the information collected from the system
+        :rtype: Dict
+        """
+
+    @abc.abstractmethod
+    def run_pba(self, name: str, options: Dict) -> None:
+        """
+        Runs a post-breach action (PBA)
+        :param str name: The name of the post-breach action to run
+        :param Dict options: A dictionary containing options that modify the behavior of the PBA
+        """
+
+    @abc.abstractmethod
+    def ping(self, host: str) -> Tuple[bool, Optional[str]]:
+        """
+        Sends a ping (ICMP packet) to a remote host
+        :param str host: The domain name or IP address of a host
+        :return: A tuple that contains whether or not the host responded and the host's inferred
+                 operating system
+        :rtype: Tuple[bool, Optional[str]]
+        """
+
+    @abc.abstractmethod
+    def scan_tcp_port(self, host: str, port: int) -> PortScanData:
```

Comment:
I believe this needs an `options` argument for `tcp_scan_timeout` and `tcp_scan_interval_ config options.

Reply:
`tcp_scan_interval` will be handled by the master. I'll add `tcp_scan_timeout`.